lib/Dialect/FIRRTL/Transforms/ResolveTraces.cpp

```diff
@@ -75,49 +75,103 @@ struct ResolveTracesPass : public ResolveTracesBase<ResolveTracesPass> {
+    };
+
+    newTarget.append("~");
+    newTarget.append(
+        path.ref.getModule()->getParentOfType<CircuitOp>().getName());
     newTarget.append("|");
 
-    if (auto nla = anno.getMember<FlatSymbolRefAttr>("circt.nonlocal")) {
-      hw::HierPathOp path = nlaTable->getNLA(nla.getAttr());
-      for (auto part : path.getNamepath().getValue().drop_back()) {
-        auto inst = cast<hw::InnerRefAttr>(part);
-        newTarget.append(inst.getModule());
-        newTarget.append("/");
-        newTarget.append(inst.getName());
-        newTarget.append(":");
-      }
+    if (path.isLocal()) {
+      addSymbol(
+          FlatSymbolRefAttr::get(path.ref.getModule().getModuleNameAttr()));
+    } else {
+      addSymbol(FlatSymbolRefAttr::get(path.instances.front()
+                                           ->getParentOfType<FModuleLike>()
+                                           .getModuleNameAttr()));
+    }
+
+    for (auto inst : path.instances) {
+      newTarget.append("/");
+      addSymbol(hw::InnerRefAttr::get(
+          inst->getParentOfType<FModuleLike>().getModuleNameAttr(),
+          inst.getInnerSymAttr().getSymName()));
+      newTarget.append(":");
+      addSymbol(inst.getModuleNameAttr());
     }
-    newTarget.append(module.getModuleName());
-    newTarget.append(">");
 
-    newTarget.append(name);
+    // If this targets a module or an instance, then we're done.  There is no
+    // "reference" part of the FIRRTL target.
+    if (path.ref.isa<OpAnnoTarget>() &&
+        path.isOpOfType<FModuleOp, FExtModuleOp, InstanceOp>())
+      return;
 
-    // Add information if this is an aggregate.
-    auto targetFieldID = anno.getFieldID();
+    newTarget.append(">");
+    auto innerSym =
+        TypeSwitch<AnnoTarget, hw::InnerSymAttr>(path.ref)
+            .Case<PortAnnoTarget>([&](PortAnnoTarget portTarget) {
+              return portTarget.getModule().getPortSymbolAttr(
+                  portTarget.getPortNo());
+            })
+            .Case<OpAnnoTarget>([&](OpAnnoTarget opTarget) {
+              return opTarget.getOp()->getAttrOfType<hw::InnerSymAttr>(
+                  "inner_sym");
+            })
+            .Default([](auto) {
+              assert(false && "unexpected annotation target type");
+              return hw::InnerSymAttr{};
+            });
+    addSymbol(hw::InnerRefAttr::get(path.ref.getModule().getModuleNameAttr(),
+                                    innerSym.getSymName()));
+
+    auto type = dyn_cast<FIRRTLBaseType>(path.ref.getType());
+    assert(type && "expected a FIRRTLBaseType");
+    auto targetFieldID = path.fieldIdx;
     while (targetFieldID) {
       FIRRTLTypeSwitch<FIRRTLBaseType>(type)
           .Case<FVectorType>([&](FVectorType vector) {
             auto index = vector.getIndexForFieldID(targetFieldID);
             newTarget.append("[");
-            newTarget.append(Twine(index).str());
+            newTarget.append(std::to_string(index));
```

Comment:
Didn't test, but doesn't look like SmallVector's append has overload for `Twine`, which makes sense.  Anyway, esp for trivial things no reason for heap traffic, consider:

```
Twine(index).toVector(newTarget);
```

(or `(Twine("[") + index + "]").toVector(newTarget);`, or similar)